### PR TITLE
Procedural map generation

### PR DESCRIPTION
**Pull request by AndreasLrx**
Generate a map randomly filled. It can use cellular automata (maybe overkill here I admit).

Resolves: #11

### CODE REVIEW

File: src/game/map/cellular/Grid.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                    for (int tempY = static_cast<int>(y) - 1; tempY < static_cast<int>(y) + 2; tempY++) {
+                        if (tempX == static_cast<int>(x) && tempY == static_cast<int>(y))
+                            continue;
+                        nbNeighbours += borderIsFilled(tempX, tempY);
+                    }
+                return nbNeighbours;
+            }
+        } // namespace cellular
+    }     // namespace map
+} // namespace bomberman
```
Comment (MisterPeModder) [inline src/game/map/cellular/Grid.cpp:98]: Missing a newline here

File: src/game/map/cellular/Grid.hpp [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+                /// Get the rule
+                ///
+                /// @return Rule rule set.
+                const Rule &getRule() const;
+
+                /// Test if the cell at the position (x, y) is filled
+                ///
+                /// @param x x position of the cell.
+                /// @param y y position of the cell.
+                /// @return true If the cell is filled.
```
Comment (MisterPeModder) [inline src/game/map/cellular/Grid.hpp:84]: Use `@retval` if you want to specify return values

File: src/game/map/cellular/Grid.hpp [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                /// @return false If the cell is empty.
+                bool borderIsFilled(int x, int y) const;
+
+                /// Get the number of filled neighbours of the cell at the position (x, y)
+                ///
+                /// @param x x position of the cell.
+                /// @param y y position of the cell.
+                /// @return unsigned char number of filled neighbours (0-8)
+                unsigned char getNeighbours(size_t x, size_t y) const;
```
Comment (MisterPeModder) [inline src/game/map/cellular/Grid.hpp:111]: We are using American English in this codebase.

File: src/game/map/cellular/Rule.cpp [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                        throw RuleInvalid("Rule with empty survive.");
+                    return 0;
+                }
+                size_t i = 1;
+                while (i < surviveRule.size() && isdigit(surviveRule[i]))
+                    _survive[surviveRule[i++] - '0'] = true;
+                return i;
+            }
+        } // namespace cellular
+    }     // namespace map
```
Comment (MisterPeModder) [inline src/game/map/cellular/Rule.cpp:84]: You have extra spaces between `}` and `// namespace`

File: src/game/map/cellular/Rule.hpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    {
+        namespace cellular
+        {
+            /// Cellular automata rule, representing of form B[0-9]+/S[0-9]+
+            /// Where B stands for Birth, S for survive and the numbers are the neighbours number
+            /// See https://en.wikipedia.org/wiki/Life-like_cellular_automaton
+            class Rule {
+              public:
+                /// Exception thrown when an error occured in the rule parsing.
+                class RuleInvalid : public std::runtime_error {
```
Comment (MisterPeModder) [inline src/game/map/cellular/Rule.hpp:28]: Do you mean `InvalidRule` ?

**Review by MisterPeModder — CHANGES_REQUESTED** [review]
Looks good, you a few things to change here and there.

**Review by ZQUEMA — APPROVED** [review]

**Pull request merged**